# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxDWriteFontList.cpp
# Commit: 095b3edec3c8
# Full Hash: 095b3edec3c8ba9318c71d0acc91e10ce57818dc
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2019-04-29 21:53:38
# Regressor Bug: 1514869
# File Overlap Count: 1
# Description:
#   Bug 1514869 - patch 2 - Adapt platform-font-list code to work with either the existing in-process font list or cross-process shared font list. r=jwatt
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D22938
# ==============================================================================

diff -r 21ef00977ab6 -r 095b3edec3c8 gfx/thebes/gfxDWriteFontList.cpp
--- a/gfx/thebes/gfxDWriteFontList.cpp	Sat Apr 27 15:37:29 2019 +0000
+++ b/gfx/thebes/gfxDWriteFontList.cpp	Sat Apr 27 15:37:58 2019 +0000
@@ -840,9 +840,10 @@
 FontFamily gfxDWriteFontList::GetDefaultFontForPlatform(
     const gfxFontStyle* aStyle) {
   // try Arial first
-  gfxFontFamily* ff;
-  if ((ff = FindFamily(NS_LITERAL_CSTRING("Arial")))) {
-    return FontFamily(ff);
+  FontFamily ff;
+  ff = FindFamily(NS_LITERAL_CSTRING("Arial"));
+  if (!ff.IsNull()) {
+    return ff;
   }
 
   // otherwise, use local default
@@ -853,12 +854,9 @@
 
   if (status) {
     ff = FindFamily(NS_ConvertUTF16toUTF8(ncm.lfMessageFont.lfFaceName));
-    if (ff) {
-      return FontFamily(ff);
-    }
   }
 
-  return FontFamily();
+  return ff;
 }
 
 gfxFontEntry* gfxDWriteFontList::LookupLocalFont(
@@ -1291,18 +1289,8 @@
   }
 }
 
-bool gfxDWriteFontList::GetStandardFamilyName(const nsCString& aFontName,
-                                              nsACString& aFamilyName) {
-  gfxFontFamily* family = FindFamily(aFontName);
-  if (family) {
-    family->LocalizedName(aFamilyName);
-    return true;
-  }
-
-  return false;
-}
-
-bool gfxDWriteFontList::FindAndAddFamilies(const nsACString& aFamily,
+bool gfxDWriteFontList::FindAndAddFamilies(StyleGenericFontFamily aGeneric,
+                                           const nsACString& aFamily,
                                            nsTArray<FamilyAndGeneric>* aOutput,
                                            FindFamiliesFlags aFlags,
                                            gfxFontStyle* aStyle,
@@ -1312,7 +1300,7 @@
 
   gfxFontFamily* ff = mFontSubstitutes.GetWeak(keyName);
   if (ff) {
-    aOutput->AppendElement(FamilyAndGeneric(ff));
+    aOutput->AppendElement(FamilyAndGeneric(ff, aGeneric));
     return true;
   }
 
@@ -1320,8 +1308,8 @@
     return false;
   }
 
-  return gfxPlatformFontList::FindAndAddFamilies(aFamily, aOutput, aFlags,
-                                                 aStyle, aDevToCssSize);
+  return gfxPlatformFontList::FindAndAddFamilies(aGeneric, aFamily, aOutput,
+                                                 aFlags, aStyle, aDevToCssSize);
 }
 
 void gfxDWriteFontList::AddSizeOfExcludingThis(MallocSizeOf aMallocSizeOf,
@@ -1472,12 +1460,20 @@
     return nullptr;
   }
 
-  gfxFontFamily* family = FindFamily(mFallbackRenderer->FallbackFamilyName());
-  if (family) {
-    gfxFontEntry* fontEntry;
-    fontEntry = family->FindFontForStyle(*aMatchStyle);
+  FontFamily family = FindFamily(mFallbackRenderer->FallbackFamilyName());
+  if (!family.IsNull()) {
+    gfxFontEntry* fontEntry = nullptr;
+    if (family.mIsShared) {
+      auto face =
+          family.mShared->FindFaceForStyle(SharedFontList(), *aMatchStyle);
+      if (face) {
+        fontEntry = GetOrCreateFontEntry(face, family.mShared);
+      }
+    } else {
+      fontEntry = family.mUnshared->FindFontForStyle(*aMatchStyle);
+    }
     if (fontEntry && fontEntry->HasCharacter(aCh)) {
-      *aMatchedFamily = FontFamily(family);
+      *aMatchedFamily = family;
       return fontEntry;
     }
     Telemetry::Accumulate(Telemetry::BAD_FALLBACK_FONT, true);